engine/crates/engine-v2/src/execution/planner/builder.rs

```diff
@@ -1,63 +1,65 @@
-use std::{borrow::Cow, collections::HashMap};
+use std::{
+    borrow::{Borrow, Cow},
+    collections::HashMap,
+};
 
+use itertools::Itertools;
 use schema::{RequiredFieldSet, ResolverWalker};
 
 use crate::{
-    execution::{ExecutableOperation, ExecutionPlan, ExecutionPlanId, PlanWalker, PreExecutionContext},
-    operation::{FieldId, LogicalPlanId, OperationWalker, SelectionSetId},
-    response::{ResponseViewSelection, ResponseViewSelectionSet, ResponseViews},
+    execution::{
+        ExecutableOperation, ExecutionPlan, ExecutionPlanId, PlanWalker, PreExecutionContext, ResponseModifierExecutor,
+    },
+    operation::{FieldId, LogicalPlanId, OperationWalker, ResponseModifierRule, SelectionSetId, SelectionSetType},
+    response::{ResponseObjectSetId, ResponseViewSelection, ResponseViewSelectionSet},
     sources::PreparedExecutor,
-    utils::BufferPool,
     Runtime,
 };
 
-use super::PlanningResult;
+use super::{BuildContext, PlanningResult};
+
+pub(super) struct ExecutionBuilder<'ctx, 'op, R: Runtime> {
+    pub(super) ctx: &'op PreExecutionContext<'ctx, R>,
+    pub(super) operation: &'op ExecutableOperation,
+    pub(super) build_context: &'op mut BuildContext,
+}
+
+impl<'ctx, 'op, R: Runtime> std::ops::Deref for ExecutionBuilder<'ctx, 'op, R> {
+    type Target = BuildContext;
+    fn deref(&self) -> &Self::Target {
+        self.build_context
+    }
+}
 
-pub(super) struct RequirementsBuildContext<'ctx, 'op, 'parent, R: Runtime> {
-    ctx: &'op PreExecutionContext<'ctx, R>,
-    operation: &'op ExecutableOperation,
-    response_views: &'parent mut ResponseViews,
-    response_view_selection_buffer_pool: &'parent mut BufferPool<ResponseViewSelection>,
+impl<'ctx, 'op, R: Runtime> std::ops::DerefMut for ExecutionBuilder<'ctx, 'op, R> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.build_context
+    }
```

Comment:
Well have to admit I'm probably abusing this deref pattern and that's not great. I'm not entirely sure how to solve it though. I often end up with basically 3 levels of context:
- `ctx`: PreExecutionContext / ExecutionContext which holds the engine, schema typically
- A root struct for the current phase: binding, solving requirements, computing shapes, building executors/execution plans
- A nested struct with additional local context: the current plan being built, the current selection set being solved, etc..

I'm always using `Deref` so that the 3rd level has access to the 2nd one and I add utility methods for schema/walker at the 2nd one usually. Without `Deref` I would need to add an extra field for like 3/4 of the `self.` usage.

But maybe I should replace the 2nd level with a Context, have its function be free-standing functions and pass that around in the 3rd level with local context

Reply:
> Without `Deref` I would need to add an extra field for like 3/4 of the `self.` usage.

Yeah, I am not sure either - one thing that occurred to me is literally just doing `let context = &self.build_context;` at the start of the function (or similarly `let Self { ctx, build_context } = &self`).  Although I don't know if lifetimes/mutability restrictions or whatever are going to fuck that up. 

> But maybe I should replace the 2nd level with a Context, have its function be free-standing functions and pass that around in the 3rd level

Yeah a few extra parameters (potentially on free standing functions) are another possibility.

Without actually trying things I don't really know the solution (and also don't think it _needs_ to be solved right now) - just wanted to raise it as a concern.  Deal with it as you like (or don't)

Reply:
I think you're right and needs to be fixed, what I'm doing is akin to the cake pattern in Scala which is an anti-pattern. And here it's particularly bad with the mutability issue of the operation leading to even more complexity. Since I'm re-using this same anti-pattern all the time, need to figure out how to structure it properly without it. I keep coming back to it being the lowest-effort solution for me right now. I'll do it next week after field analytics